### PR TITLE
GUI Game Lobby

### PR DESCRIPTION
**Pull request by TimTheToad**
- Can connect via IP (keep in mind Server needs to run as well)
- Design Overhaul
- Saved Server List(WIP) - Will later allow users to save Previously played Servers as "Favorites"
- Local Server Browser(WIP) - Will later allow users to see servers that are up and running on your local network.
- Host Server(WIP) - Users may now host a server - They may choose map & player count. 

Many of these things are currently WIP and dependant on things on the Networking side but to keep the master updated and not falling too far behind from master a pull request is made.

### CODE REVIEW

File: CrazyCanvas/Include/GUI/SavedServerGUI.h [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+	void AddServerItem(Noesis::Grid* pParentGrid, const char* pServerN, const char* pMapN, const char* pPing, bool isRunning);
+
+    void Init(Noesis::ListBox* pListView);
+
+    
+private:
+	uint8 m_ItemCount;
+
+    Noesis::Ptr<Noesis::ListBox> m_pSavedServerList;
```
Comment (Vonsen) [inline CrazyCanvas/Include/GUI/SavedServerGUI.h]:
```suggestion
    Noesis::Ptr<Noesis::ListBox> m_SavedServerList;
```
Isn't that a shared pointer? Mind is a bit fuzzled but I thought we agreed not to use p in these cases.

File: CrazyCanvas/Source/GUI/LobbyGUI.cpp [hunk 2]
```diff
@@ -1,25 +1,56 @@
...
 
-LobbyGUI::LobbyGUI(const LambdaEngine::String& xamlFile)
+LobbyGUI::LobbyGUI(const LambdaEngine::String& xamlFile) :
+	m_HostGameDesc(),
+	m_ServerList(xamlFile)
 {
 	Noesis::GUI::LoadComponent(this, xamlFile.c_str());
+	//m_pRoot = Noesis::GUI::LoadXaml<Grid>(xamlFile.c_str());
+
+	const char* ip = "192.168.1.65";
```
Comment (Vonsen) [inline CrazyCanvas/Source/GUI/LobbyGUI.cpp]:
```suggestion
	const char* pIP = "192.168.1.65";
```

File: CrazyCanvas/Source/GUI/LobbyGUI.cpp [hunk 3]
```diff
@@ -1,25 +1,56 @@
...
+	m_ServerList.Init(FrameworkElement::FindName<ListBox>("SAVED_SERVER_LIST"));
+	ErrorPopUpClose();
 }
 
 LobbyGUI::~LobbyGUI()
 {
+
 }
 
 bool LobbyGUI::ConnectEvent(Noesis::BaseComponent* source, const char* event, const char* handler)
```
Comment (Vonsen) [inline CrazyCanvas/Source/GUI/LobbyGUI.cpp:46]:
```suggestion
bool LobbyGUI::ConnectEvent(Noesis::BaseComponent* pSource, const char* pEvent, const char* pHandler)
```

File: CrazyCanvas/Source/GUI/LobbyGUI.cpp [hunk 4]
```diff
@@ -29,3 +60,127 @@ void LobbyGUI::OnButtonBackClick(Noesis::BaseComponent* pSender, const Noesis::R
 	StateManager::GetInstance()->EnqueueStateTransition(pMainMenuState, STATE_TRANSITION::POP_AND_PUSH);
 }
 
+void LobbyGUI::OnButtonConnectClick(Noesis::BaseComponent* pSender, const Noesis::RoutedEventArgs& args)
+{
+	LOG_MESSAGE(FrameworkElement::FindName<TextBox>("IP_ADDRESS")->GetText());
+
+	IPAddress* ip = IPAddress::Get(FrameworkElement::FindName<TextBox>("IP_ADDRESS")->GetText());
```
Comment (Vonsen) [inline CrazyCanvas/Source/GUI/LobbyGUI.cpp]:
```suggestion
	IPAddress* pIP = IPAddress::Get(FrameworkElement::FindName<TextBox>("IP_ADDRESS")->GetText());
```

File: CrazyCanvas/Source/GUI/LobbyGUI.cpp [hunk 5]
```diff
@@ -29,3 +60,127 @@ void LobbyGUI::OnButtonBackClick(Noesis::BaseComponent* pSender, const Noesis::R
...
+	RenderSystem::GetInstance().SetRenderStageSleeping("RENDER_STAGE_NOESIS_GUI",	true);
+
+	/*if (m_RayTracingEnabled)
+		RenderSystem::GetInstance().SetRenderStageSleeping("RAY_TRACING", m_RayTracingSleeping);*/
+
+}
+
+void LobbyGUI::ErrorPopUp(ErrorCode errorCode)
+{
+	TextBlock* textBox = FrameworkElement::FindName<TextBlock>("ERROR_BOX_TEXT");
```
Comment (Vonsen) [inline CrazyCanvas/Source/GUI/LobbyGUI.cpp]:
```suggestion
	TextBlock* pTextBox = FrameworkElement::FindName<TextBlock>("ERROR_BOX_TEXT");
```

File: CrazyCanvas/Source/GUI/SavedServerGUI.cpp [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+void SavedServerGUI::AddServerItem(Grid* pParentGrid, const char* pServerN, const char* pMapN, const char* pPing, bool isRunning)
+{
+	
+	Ptr<Grid> grid = *new Grid();
+
+	for (int i = 0; i < 4; i++)
+	{
+		Ptr<ColumnDefinition> pColumnDef = *new ColumnDefinition();
```
Comment (Vonsen) [inline CrazyCanvas/Source/GUI/SavedServerGUI.cpp]:
```suggestion
		Ptr<ColumnDefinition> columnDef = *new ColumnDefinition();
```
Shared pointer? View first comment

File: CrazyCanvas/Source/GUI/SavedServerGUI.cpp [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+		GridLength gridLength = GridLength(25, GridUnitType_Star);
+		pColumnDef->SetWidth(gridLength);
+		grid->GetColumnDefinitions()->Add(pColumnDef);
+	}
+
+	Ptr<RowDefinition> pRowDef = *new RowDefinition();
```
Comment (Vonsen) [inline CrazyCanvas/Source/GUI/SavedServerGUI.cpp]:
```suggestion
	Ptr<RowDefinition> rowDef = *new RowDefinition();
```

File: CrazyCanvas/Source/GUI/SavedServerGUI.cpp [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+	//m_pSavedServerList->GetChildren()->Add(m_pSavedServerList);
+	//gl�m inte s�tta row ocks� n�r du l�st detta bajsproblemet. 
+	pParentGrid->SetColumn(grid, 1);
+	pParentGrid->SetColumnSpan(grid, 2);
+	pParentGrid->SetRow(grid, 4);
+
+	Ptr<TextBlock> pServerName	= *new TextBlock();
+	Ptr<TextBlock> pMapName		= *new TextBlock();
+	Ptr<TextBlock> pping		= *new TextBlock();
+	Ptr<Rectangle> pIsRunning	= *new Rectangle();
```
Comment (Vonsen) [inline CrazyCanvas/Source/GUI/SavedServerGUI.cpp]:
```suggestion
	Ptr<TextBlock> serverName	= *new TextBlock();
	Ptr<TextBlock> mapName		= *new TextBlock();
	Ptr<TextBlock> ping		= *new TextBlock();
	Ptr<Rectangle> isRunning	= *new Rectangle();
```

File: CrazyCanvas/Source/GUI/SavedServerGUI.cpp [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+
+	pServerName->SetText(pServerN);
+	pMapName->SetText(pMapN);
+	pping->SetText(pPing);
+
+	Ptr<SolidColorBrush> pBrush = *new SolidColorBrush();
```
Comment (Vonsen) [inline CrazyCanvas/Source/GUI/SavedServerGUI.cpp]:
```suggestion
	Ptr<SolidColorBrush> brush = *new SolidColorBrush();
```

File: CrazyCanvas/Source/GUI/LobbyGUI.cpp (hunk 3 above)
Comment (TimTheToad) [reply CrazyCanvas/Source/GUI/LobbyGUI.cpp:46]: Not Done

**Review by Vonsen — APPROVED** [review]

**Pull request merged**